### PR TITLE
92/use order and erc20s hook

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Part of #92
Waterfalls into https://github.com/gnosis/gp-ui/pull/146/files
Follow up to https://github.com/gnosis/gp-ui/pull/146#discussion_r576417353

New order type, transforming between `RawOrder` and `Order` (the new type) and new hook vastly improving `OrderWidget`.

Still more refactor to be done. We'll get there.

### CODE REVIEW

File: src/api/operator/types.ts [hunk 1]
```diff
@@ -38,6 +40,30 @@ export type RawOrder = {
...
+  buyTokenAddress: string
+  sellTokenAddress: string
+  buyAmount: BigNumber
+  sellAmount: BigNumber
+  executedBuyAmount: BigNumber
+  executedSellAmount: BigNumber
+  feeAmount: BigNumber
+  executedFeeAmount: BigNumber
+  cancelled: boolean
+  // status: OrderStatus // better not have the status because it depends on `expirationDate`
```
Comment (W3stside) [inline src/api/operator/types.ts:60]: are you keeping these commented props for a reason

Comment (alfetopito) [reply src/api/operator/types.ts:60]: So we can talk about it :)
Will remove

File: src/api/operator/types.ts [hunk 2]
```diff
@@ -38,6 +40,28 @@ export type RawOrder = {
   signature: string
 }
 
+/**
+ * Enriched Order type.
+ * Applies some transformations on the raw api data.
+ * Some fields are kept as is.
+ */
+export type Order = Pick<RawOrder, 'owner' | 'uid' | 'appData' | 'kind' | 'partiallyFillable' | 'signature'> & {
```
Comment (anxolin) [inline src/api/operator/types.ts:48]: 🍝😋 delicious

File: src/api/operator/types.ts [hunk 3]
```diff
@@ -38,6 +40,28 @@ export type RawOrder = {
...
+  creationDate: Date
+  expirationDate: Date
+  buyTokenAddress: string
+  sellTokenAddress: string
```
Comment (anxolin) [inline src/api/operator/types.ts:52]: if we enrich this, should we also enrich the token? not sure, just thinking...

File: src/apps/explorer/components/OrderWidget/index.tsx [hunk 4]
```diff
@@ -1,62 +1,24 @@
...
-    addresses: [order?.buyToken || '', order?.sellToken || ''].filter(Boolean),
-    networkId: networkIdOrDefault,
-  })
-  const [buyToken, sellToken] = useMemo(() => {
-    const buyToken = erc20s[order?.buyToken || ''] || undefined
-    const sellToken = erc20s[order?.sellToken || ''] || undefined
-
-    return [buyToken, sellToken]
-  }, [order?.buyToken, order?.sellToken, erc20s])
+  const { order, buyToken, sellToken, isLoading, error } = useOrderAndErc20s(orderId, ORDER_QUERY_INTERVAL)
```
Comment (anxolin) [inline src/apps/explorer/components/OrderWidget/index.tsx:13]: it feels right when you can kill this much logic from the component

Comment (anxolin) [inline src/apps/explorer/components/OrderWidget/index.tsx:13]: Related to my comment above. This hooks looks fairly easy to use. Although one could think that buyToken, sellToken are part of the order. So ideally, the use would be:
```
const { order, isLoading, error } = useOrderAndErc20s(orderId, ORDER_QUERY_INTERVAL)

order.sellTokenAddress // address, we can leave it or kill this prop since it's derived from the sellToken
order.sellToken // token
```

File: src/components/orders/OrderDetails/index.tsx [hunk 5]
```diff
@@ -91,11 +99,11 @@ export function OrderDetails(props: Props): JSX.Element {
           </tr>
           <tr>
             <td>Sell amount</td>
-            <td>{`${formatSmart(sellAmount, sellToken.decimals)} ${sellToken.symbol}`}</td>
+            <td>{`${formatSmart(sellAmount.toString(), sellToken.decimals)} ${sellToken.symbol}`}</td>
```
Comment (anxolin) [inline src/components/orders/OrderDetails/index.tsx:102]: ubernit. not important cause it's always hard to make changes to dex-js, and this one is too little. But if we touch it any time soon, maybe we want BigInt support (to save us the toString every time we format ---> no deal at all)

File: src/hooks/useOperatorOrder.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  isLoading: boolean
+  forceUpdate: () => void
+}
+
+export function useOrder(orderId: string): UseOrderResult {
+  const [isLoading, setIsLoading] = useState(false)
+  const [error, setError] = useState('')
+  const [order, setOrder] = useState<Order | null>(null)
+  // Hack to force component to update itself on demand
+  const [forcedUpdate, setForcedUpdate] = useState({})
```
Comment (anxolin) [inline src/hooks/useOperatorOrder.ts:23]: I don't see where you use this

File: src/hooks/useOperatorOrder.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ * @param orderId The order id
+ * @param updateInterval [Optional] How often should try to update the order
+ */
+export function useOrderAndErc20s(orderId: string, updateInterval = 0): UseOrderAndErc20sResult {
+  const networkId = useNetworkId() || undefined
+
+  const { order, isLoading: isOrderLoading, error: orderError, forceUpdate } = useOrder(orderId)
+  const addresses = order ? [order.buyTokenAddress, order.sellTokenAddress] : []
+
+  // TODO: consume errors obj
```
Comment (anxolin) [inline src/hooks/useOperatorOrder.ts:76]: u leave this one for another PR?

Maybe is just exposing it directly --> return erc20Errors also

File: src/hooks/useOperatorOrder.ts (hunk 6 above)
Comment (anxolin) [reply src/hooks/useOperatorOrder.ts:23]: EDiT: All right, I see you call this from the other useOrder...

File: src/hooks/useOperatorOrder.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+  useEffect(() => {
+    let intervalId: NodeJS.Timeout | null = null
+
+    // Only start the interval when:
+    // 1. Hook is configured to do so (`updateInterval` > 0)
+    // 2. Order exists
+    // 3. Order is not expired
+    if (updateInterval && order && order.expirationDate.getTime() > Date.now()) {
+      intervalId = setInterval(forceUpdate, updateInterval)
```
Comment (anxolin) [inline src/hooks/useOperatorOrder.ts:90]: Shouldn't this logic be on useOrder already? so you pass that hook the `updateInterval`

This way, you have the 4 options of use:
- basic order --> interval = 0
- basic order --> interval = 20s (for example)
- detailed order --> interval = 0
- detailed order --> interval = 20s (for example)

**Review by anxolin — COMMENTED** [review]
Great work! mostly approved, just some discussions, but I'm happy with the current approach

File: src/api/operator/types.ts (hunk 3 above)
Comment (alfetopito) [reply src/api/operator/types.ts:52]: I though about this.
The reasons for keeping it out of the `Order` are:
1. Token details are not part of the raw order, thus will have to be added later
2. Avoid nested objects so React is better at identifying obj updates/component redraw

For `1`, it's easy to address, since I'm fetching the token details in the same hook
For `2`, haven't really measured the performance impact. Was more of something I'm aware but don't fully know how it works.

None are strong reasons though. Will update.

File: src/apps/explorer/components/OrderWidget/index.tsx (hunk 4 above)
Comment (alfetopito) [reply src/apps/explorer/components/OrderWidget/index.tsx:13]: It was all temporary throw away code anyway :)

File: src/components/orders/OrderDetails/index.tsx (hunk 5 above)
Comment (alfetopito) [reply src/components/orders/OrderDetails/index.tsx:102]: Agreed. https://github.com/gnosis/gp-ui/issues/179

File: src/hooks/useOperatorOrder.ts (hunk 7 above)
Comment (alfetopito) [reply src/hooks/useOperatorOrder.ts:76]: Yeah, should do something about it.
The issue is that this error is a mapping of address -> error msg

Think the best thing to do is to merge the order error with this obj and return the obj instead of a string.

File: src/hooks/useOperatorOrder.ts (hunk 8 above)
Comment (alfetopito) [reply src/hooks/useOperatorOrder.ts:90]: Right, makes sense to move the update logic there.
I just don't get your 4 options. What do you mean?

File: src/api/operator/types.ts (hunk 2 above)
Comment (alfetopito) [reply src/api/operator/types.ts:48]: :pie:

File: src/hooks/useOperatorOrder.ts (hunk 8 above)
Comment (alfetopito) [reply src/hooks/useOperatorOrder.ts]: Updated and still don't get your options :D

File: src/api/operator/types.ts (hunk 3 above)
Comment (anxolin) [reply src/api/operator/types.ts:55]: It was ok too. I didn't have strong opinion on this one either. I see reasons to do it and not do it

File: src/hooks/useOperatorOrder.ts (hunk 8 above)
Comment (anxolin) [reply src/hooks/useOperatorOrder.ts]: ahaha, does't matter :) 

I was just showing that if you move it to the basic order hook, then anyone can use it as it is today, just pass a 0 to the interval. And then allow to set an interval. So it has both uses.

And the signature of the detailed one doesn't change. So it's win win.

**Review by anxolin — APPROVED** [review]

File: src/apps/explorer/components/OrderWidget/view.tsx [hunk 9]
```diff
@@ -14,27 +12,26 @@ const Wrapper = styled.div`
...
     <Wrapper>
       <h2>Order details</h2>
       {/* TODO: create common loading indicator */}
-      {isLoading && <FontAwesomeIcon icon={faSpinner} spin size="5x" />}
-      {order && !isLoading && buyToken && sellToken && (
-        <OrderDetails order={order} buyToken={buyToken} sellToken={sellToken} />
-      )}
+      {order?.buyToken && order?.sellToken && <OrderDetails order={order} />}
       {!order && !isLoading && <p>Order not found</p>}
+      {!isLoading && order && (!order?.buyToken || !order?.sellToken) && <p>Not able to load tokens</p>}
```
Comment (W3stside) [inline src/apps/explorer/components/OrderWidget/view.tsx:29]: don't think you need `order` here if you use the `nullish coalesence` of `!order?.buToken` and sell

**Review by W3stside — COMMENTED** [review]
approve

File: src/apps/explorer/components/OrderWidget/view.tsx (hunk 9 above)
Comment (alfetopito) [reply src/apps/explorer/components/OrderWidget/view.tsx:29]: I actually need it. Only want to show it when one of the tokens failed loading.
Otherwise when the order is not found both lines 28 and 29 will be rendered.
